### DISCUSSION TITLE
An option to disable registration in self-hosted setup

### DISCUSSION THREAD

**Discussion by thelazyoxymoron** [solidtime-io/solidtime#235] [Feature requests]
Would love to have an option to disable user registrations

**Comment by Onatcer** [comment]
Hi, thank you for the feature request. 

That's a good point, we'll look into that. I'm not quite sure how we can best handle the initial user registration, as well as adding new users, in that case. 

My initial thoughts: We could either do a setting in the admin panel somewhere, or it would be an environment variable. 

If anyone has input feel free to add it in the comments to this issue.

**Comment by thelazyoxymoron** [comment]
Environment variable should be sufficient in this case

**Comment by micalm** [comment]
I agree that an env would be enough; this is how [Gitea](https://docs.gitea.com/administration/config-cheat-sheet#service-service) does it.

They do have an registration (or rather user creation) view for the admin, which would require additional consideration for solidtime:
1. Do users get sent a password via email or do they have to set it themselves?
2. Are personal organizations created or are users forced to live in a certain organization?
3. Something else I'm probably missing now

---

Another approach that I think would be similar in effort required is a Slack-like `ALLOWED_EMAILS` or something, where guests **can** register, as long as they have an email in a domain allowed by the instance admin. This doesn't require any changes or new flows except a small message that says `you can't register from that email address, bye`.

**Comment by thelazyoxymoron** [comment]
Isn't there already a "new user" creation functionality for admins in Solidtime as well? I can see one in `/admin/users` page. Or am I missing something?

**Comment by Onatcer** [comment]
@thelazyoxymoron yes you can create users, but I'm pretty sure you can break the application with that right now, as in our regular flow you'll always have a organization as well, and basically everything happens in the organization. Not sure if you can assign users to organizations via the admin panel right now and if email confirmation stuff works correctly in that case. 

But the bigger blocker is that you need to register an account to access the admin panel first, so we have to create some flow for that to set up an initial user.

**Comment by korridor** [comment]
We decided to move feature requests into github discussions, to separate feature requests from bugs and allow for better prioritization. If this feature is important to you, you can upvote it to let us know.

**Comment by kosli** [comment]
I was also looking into a way to disable the user registration, as I want to use solidtime with my "closed" user group and not have random people (or bots) signing up on the publicly available instance. An env variable would be definitely sufficient, whereas new users could be invited from the organization panel.

**Comment by dallyger** [comment]
> I'm not quite sure how we can best handle the initial user registration

I think, that that is not actually a blocker. A viable work-around would be to create the instance, register as an user and then set the env variable or the option to disable any new registrations.

Of course, having an actual flow for that would be great. Maybe as a cli command like `create:user --admin`? But my guess is, that most people would be fine with the work-around.

**Comment by korridor** [comment]
We just pushed a feature into the `main` that allows you to disable/enable the registration with an environment variable called `APP_ENABLE_REGISTRATION`. The values are `true` or `false`. Per default the registration will be deactivated. We'll add a warning in the next official release, regarding this change.

It would be great if someone could test if this feature works as expected, before we officially release the change.

We also added a new artisan command `admin:user:create` to create users manually via the CLI.

**Comment by korridor** [comment]
✅ This feature is now available in version [0.5.0](https://github.com/solidtime-io/solidtime/releases/tag/v0.5.0).

See [last comment](https://github.com/solidtime-io/solidtime/discussions/235#discussioncomment-12099406) for more information

**Comment by henningtillmann** [comment]
I have updated to the new version and set `APP_ENABLE_REGISTRATION="false"`. The registration screen is still accessible.

**Reply by korridor** [reply]
Yes, but you will get an error saying that the registration is disabled if you try to submit the form. It is not the most beautiful solution, but I think it will suffice for most setups for now.

**Comment by Eiernase** [comment]
I just noticed that when this variable is set to false invites don't work.
Maybe allow registration for invited users?
My workaround right now is to create a user first and then add them to the organization I want to invite them to, but that creates a default personal organization for them first on account creation.

Or, you know, implement OIDC :D

**Reply by gillesgw** [reply]
Hi everyone, same here, for adding a member via invite I must set the variable, let the new user create account, and then set false again; I think APP_ENABLE_REGISTRATION=false should not impact invites.
